### ISSUE TITLE
Improving CI/CD test coverage

### ISSUE THREAD

**Issue by rafael** [vitessio/vitess#4989]
# Expanding Vitess Test Coverage 

Currently Vitess has a rich set of functional tests that are run as part of every commit to catch regressions early.  However, they are not sufficient to assess the quality of the product for production rollout.  

Some of the problems with the current approach are:

- The impact of a commit on integration scenarios is not caught in a timely manner, and in case of a regression is leaving the code in an unhealthy state. Without a timely detection of this regression, additional commits continue to land on an already unhealthy codebase.
- The effort to integrate with the latest codebase is taking longer due to the unknown state of the codebase and uncaught regressions, performance, compatibility issues.

### Continuous Integration:
**Frequency**: For every commit
**Scope**: Acceptance and Unit Testing

_Note: This what we already have by using travis._

### Regression Testing

**Frequency**: Nightly
**Scope**: Functionality, Reliability, Backward/Forward compatibility

Automated test that runs system level checks including spinning up a cluster with the build, setting up some data, making sure basic things work, upgrading the various components, ensuring things still work, downgrading back to the previous build, re-validating things still work, etc. Operations that should be tested in this flow: 

- Backups: 
- They work with the new  build.
- PlannedReparents
- Restores
- Tablets backed up with a new build can be restored using the previous build. 
- Tablets backed up with previous build can be restored using new build.
- V3 Routing.

### Production Readiness

**Frequency:**  Pick a new build every 2 weeks
**Scope:** Performance, Stress

Automated test that tracks overall performance of Vitess components over time. 

- Horizontal resharding an Shard merges
- Write/Read Throughput for Vtgates/Tablets
- Wriite/Read latency  for Vtgates/Tablets

**Comment by morgo** [comment]
I like it! I just have one suggestion: under regression testing, we should test all MySQL flavors and versions Vitess claims to support.

**Comment by rafael** [comment]
Oh that's a great idea. Just added that!

**Comment by derekperkins** [comment]
This is great. I think we should also add release tagging as a part of the production readiness cadence.

**Comment by morgo** [comment]
If everyone is okay with it, I would like to take a look at this after I've finished the release cycle documentation + small documentation refactoring tasks I'm working on (~1-2 weeks time)

**Comment by rafael** [comment]
@morgo  yes! I think that makes sense. I see the release cycle/doc as a pre-req for the improvements in this issue.

**Comment by morgo** [comment]
I am going to look at Circle CI & AWS Code Build in scope as well. @dkhenry suggested that if we parallelize the tests more we can run the regression suite on every commit (versus nightly).

That makes sense to me. We can revert the plan to nightly if the cost/time is prohibitive.

**Comment by morgo** [comment]
I am planning to loop back on this. I just want to merge a couple of PRs that change the build/testing environment:

* https://github.com/vitessio/vitess/pull/5382 
* https://github.com/vitessio/vitess/pull/5326

Work is also underway to remove python from the tests. @arindamnayak, @ajeetj and @saurabh408 are all working on it :-)

What I plan to do is first move `local_example` to use GitHub actions as a true matrix build on supported flavors; since this is python-less. Assuming we can run tests on new infrastructure much faster, we can go with the "everything on commit" plan, which simplifies having to think about things.

**Comment by aribalam** [comment]
@GuptaManan100 I am willing to work on this project. The way I understand, there are 2 tasks.
1. All the functional tests that are run in Travis CI need to be migrated to Github actions.
2. We need to create scenarios that requires users to downgrade a level in case of a failed upgrade. Moreover, we need to create Github Actions that will enable us to checkout 2 versions for testing.

Is there anything that needs to be known to better understand the project?

**Comment by GuptaManan100** [comment]
The mentor for this project is going to be @harshit-gangal. He will be best able to answer your queries.

**Comment by harshit-gangal** [comment]
The Vitess Test Suite runs on Github Actions today, there is no Travis CI anymore.
We need to test the real-world scenario of upgrading to a newer version in any order (Vttablet or Vtgate first) and then downgrading while serving traffic.

**Comment by aribalam** [comment]
@harshit-gangal Great! :)
Could you give me any starting lead on how to get started so as to understand the project better?

**Comment by deepthi** [comment]
This might help: https://github.com/vitessio/vitess/issues/7344

**Comment by aribalam** [comment]
Okay so I went through #7344, and observed that only 2 of the scenarios still remains to be tested
1. Upgrade/downgrade subset of vttablets.
2. Upgrade/downgrade vtgate, vtctld with 1.

Also, it mentions that there needs to be some more explicit testing for vtgate and vtctld in the end-to-end tests itself.
Is there anything else that I failed to mention?
@deepthi @harshit-gangal
